File: code/raytracer/python/mesh.py
import sys
from linearAlgebra import *


class Face:
    def __init__(self, position_indices, tex_coord_indices=None, normal_indices=None, mesh=None):
        self.position_indices = position_indices
        self.tex_coord_indices = tex_coord_indices
        self.normal_indices = normal_indices
        self.mesh = mesh
        self.materials = mesh.materials
        self.face_normal = normalize(cross(sub(self.position(2), self.position(1)),
                                     sub(self.position(0), self.position(1))))
        self.barycentric_cache = None

    def coord(self, vertex, axis):
        return self.mesh.positions[self.position_indices[vertex]][axis]

    def position(self, vertex):
        return self.mesh.positions[self.position_indices[vertex]]

    def normal(self, vertex):
        if self.normal_indices is None or len(self.normal_indices) == 0:
            return self.face_normal
        return self.mesh.normals[self.normal_indices[vertex]]

    def get_normal(self, p):
        if self.barycentric_cache is None:
            assert False
        na = mult_scalar(self.barycentric_cache[0], self.normal(0))
        nb = mult_scalar(self.barycentric_cache[1], self.normal(1))
        nc = mult_scalar(self.barycentric_cache[2], self.normal(2))
        n = add(na, add(nb, nc))
        return normalize(n)

    def intersect(self, ray):
        xa = self.coord(0, 0)
        xb = self.coord(1, 0)
        xc = self.coord(2, 0)
        ya = self.coord(0, 1)
        yb = self.coord(1, 1)
        yc = self.coord(2, 1)
        za = self.coord(0, 2)
        zb = self.coord(1, 2)
        zc = self.coord(2, 2)
        xd = ray.direction[0]
        yd = ray.direction[1]
        zd = ray.direction[2]
        xe = ray.position[0]
        ye = ray.position[1]
        ze = ray.position[2]

        tMin = float("inf")

        detA = det(xa - xb, xa - xc, xd, ya - yb, ya - yc, yd, za - zb, za - zc, zd)
        t = det(xa - xb, xa - xc, xa - xe, ya - yb, ya - yc, ya - ye, za - zb, za - zc,
                za - ze) / detA
        if t < 0:
            return tMin, None
        gamma = det(xa - xb, xa - xe, xd, ya - yb, ya - ye, yd, za - zb, za - ze, zd) / detA
        if gamma < 0 or gamma > 1:
            return tMin, None
        beta = det(xa - xe, xa - xc, xd, ya - ye, ya - yc, yd, za - ze, za - zc, zd) / detA
        if beta < 0 or beta > (1 - gamma):
            return tMin, None
        tMin = t
        self.barycentric_cache = [1 - beta - gamma, beta, gamma]
        return tMin, self


class Mesh:
    def __init__(self, file_path=None, materials=None):
        self.file_path = file_path
        self.materials = materials

        self.positions = []
        self.normals = []
        self.tex_coordinates = []
        self.faces = []
        self.bounding_box = None

    def init(self):
        if self.file_path is not None:
            self._load_from_obj(self.file_path)
            self._create_bounding_box()

    def _create_bounding_box(self):
        self.bounding_box = Mesh()
        if len(self.positions) == 0:
            return
        min = [float("inf"), float("inf"), float("inf")]
        max = [float("-inf"), float("-inf"), float("-inf")]
        for p in self.positions:
            for i, v in enumerate(p):
                if v < min[i]:
                    min[i] = v
                if v > max[i]:
                    max[i] = v

        a, b = min, max
        self.bounding_box.positions = [
                     [a[0], a[1], a[2]], [a[0], a[1], b[2]], [a[0], b[1], a[2]],
                     [a[0], b[1], b[2]], [b[0], a[1], a[2]], [b[0], a[1], b[2]],
                     [b[0], b[1], a[2]], [b[0], b[1], b[2]]
                     ]
        self.bounding_box.faces = [
                                   Face([0, 2, 1], mesh=self.bounding_box),
                                   Face([1, 2, 3], mesh=self.bounding_box),
                                   Face([4, 5, 6], mesh=self.bounding_box),
                                   Face([5, 7, 6], mesh=self.bounding_box),
                                   Face([0, 1, 4], mesh=self.bounding_box),
                                   Face([4, 1, 5], mesh=self.bounding_box),
                                   Face([2, 6, 3], mesh=self.bounding_box),
                                   Face([3, 6, 7], mesh=self.bounding_box),
                                   Face([0, 4, 2], mesh=self.bounding_box),
                                   Face([4, 6, 2], mesh=self.bounding_box),
                                   Face([1, 3, 5], mesh=self.bounding_box),
                                   Face([3, 7, 5], mesh=self.bounding_box)
                                   ]
        self.bounding_box.mesh = self.bounding_box

    def _load_from_obj(self, file_path):
        with open(file_path) as f:
            content = f.readlines()
            for line in content:
                parts = line.split()
                if len(parts) == 0:
                    continue
                parts = [p.strip() for p in parts]
                if parts[0] == 'v':
                    self.positions.append([float(parts[i]) for i in range(1, 4)])
                elif parts[0] == 'vt':
                    self.tex_coordinates.append([float(parts[i]) for i in range(1, 3)])
                elif parts[0] == 'vn':
                    self.normals.append([float(parts[i]) for i in range(1, 4)])
                elif parts[0] == 'f':
                    position_indices = []
                    tex_coord_indices = []
                    normal_indices = []
                    for part in [parts[i] for i in range(1,4)]:
                        indices = part.split("/")
                        if len(indices) > 0 and indices[0] != "":
                            position_indices.append(int(indices[0]) - 1)
                        if len(indices) > 1 and indices[1] != "":
                            tex_coord_indices.append(int(indices[1]) - 1)
                        if len(indices) > 2 and indices[2] != "":
                            normal_indices.append(int(indices[2]) - 1)
                    self.faces.append(Face(position_indices, tex_coord_indices,
                                           normal_indices, mesh=self))

    def intersect(self, ray):
        if self.bounding_box is not None:
            tMin, _ = self.bounding_box.intersect(ray)
            if tMin == float("inf"):
                return float("inf"), None
        tMin = float("inf")
        intersectedFace = None
        for face in self.faces:
            t, f = face.intersect(ray)
            if t < tMin:
                tMin = t
                intersectedFace = f
        return tMin, intersectedFace


if __name__ == '__main__':
    file_path = 'sphere.obj'
    if len(sys.argv) > 1:
        file_path = sys.argv[1]
    mesh = Mesh()
    mesh.load_from_obj(file_path)

File: code/raytracer/c#/raytracer/Mesh.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace raytracer
{
  class Mesh : IObject
  {
    private readonly bool _computeVertexNormals;
    private Transform _transform;
    private Box _boundingBox;
    private OctTreeNode _octTree;

    public string FilePath { get; set; }
    public List<string> Materials { get; set; }
    
    public List<Vector> Positions { get; private set; } 
    public List<Vector> Normals { get; private set; } 
    public List<Vector> TexCoords { get; private set; } 
    public List<Face> Faces { get; private set; } 

    public Mesh(string filePath, List<string> materials, bool computeVertexNormals, Transform transform)
    {
      _computeVertexNormals = computeVertexNormals;
      _transform = transform;
      _boundingBox = new Box(new Vector(float.MaxValue, float.MaxValue, float.MaxValue), new Vector(float.MinValue, float.MinValue, float.MinValue));

      FilePath = filePath;
      Materials = materials;

      Positions = new List<Vector>();
      Normals = new List<Vector>();
      TexCoords = new List<Vector>();
      Faces = new List<Face>();
    }

    public void Init()
    {
      LoadFromObj(FilePath);
      _octTree = OctTreeNode.GenerateOctTree(_boundingBox, 0, 3);
      _octTree.Initialize(this);
      if (_computeVertexNormals)
        ComputeVertexNormals();
    }

    private void LoadFromObj(string filePath)
    {
      using (var reader = new StreamReader(filePath))
      {
        while (!reader.EndOfStream)
        {
          var line = reader.ReadLine();
          var parts = line.Split();
          if (parts.Length == 0)
            continue;

          if (parts[0] == "v")
          {
            var vp = new Vector(float.Parse(parts[1]), float.Parse(parts[2]), float.Parse(parts[3]));
            var tvp = _transform.TransformPosition(vp);
            Positions.Add(tvp);
            _boundingBox.Update(tvp);
          }
          else if (parts[0] == "vt")
            TexCoords.Add(new Vector(float.Parse(parts[1]), float.Parse(parts[2]), 0));
          else if (parts[0] == "vn")
          {
            var vn = new Vector(float.Parse(parts[1]), float.Parse(parts[2]), float.Parse(parts[3]));
            Normals.Add(_transform.TransformDirection(vn).Normalized);
          }
          else if (parts[0] == "f")
          {
            ParseFace(parts);
          }
        }
      }
    }

    private void ParseFace(string[] parts)
    {
      var positionIndices = new List<int>();
      var texCoordIndices = new List<int>();
      var normalIndices = new List<int>();
      for (int i = 1; i < 4; i++)
      {
        var indices = parts[i].Split(new[] { '/' }, StringSplitOptions.None);
        if (indices.Length > 0 && indices[0] != "")
          positionIndices.Add(int.Parse(indices[0]) - 1);
        if (indices.Length > 1 && indices[1] != "")
          texCoordIndices.Add(int.Parse(indices[1]) - 1);
        if (indices.Length > 2 && indices[2] != "")
          normalIndices.Add(int.Parse(indices[2]) - 1);
      }
      Faces.Add(new Face(positionIndices, texCoordIndices, normalIndices, this));
    }

    private void ComputeVertexNormals()
    {
      if (Normals.Count > 0)
        return;
      List<List<Face>> facesForVertices = GetFacesForVertices();
      foreach (var facesForVertex in facesForVertices)
      {
        Vector normal = facesForVertex.Select(f => f.FaceNormal).Aggregate((acc, v) => acc + v).Normalized;
        Normals.Add(normal.Normalized);
      }
      foreach (var face in Faces)
      {
        face.NormalIndices = face.PositionIndices.ToList();
      }
    }

    private List<List<Face>> GetFacesForVertices()
    {
      var facesForVertices = new List<List<Face>>();
      foreach (var position in Positions)
      {
        facesForVertices.Add(new List<Face>());
      }
      foreach (var face in Faces)
      {
        foreach (var positionIndex in face.PositionIndices)
        {
          facesForVertices[positionIndex].Add(face);
        }
      }

      return facesForVertices;
    }

    public Tuple<float, IObject> Intersect(Ray ray)
    {
      return _octTree.Intersect(ray);
    }

    public Vector GetNormal(Vector p, float time)
    {
      return new Vector();
    }

    public Vector GetTextureCoords(Vector p, float time)
    {
      return new Vector();
    }
  }
}
